app/src/main/java/org/thoughtcrime/securesms/blurhash/BlurHashEncoder.java

```diff
@@ -0,0 +1,148 @@
+    return encode(bitmap, 4, 3);
+  }
+
+  static String encode(Bitmap bitmap, int componentX, int componentY) {
+    int width    = bitmap.getWidth();
+    int height   = bitmap.getHeight();
+    int[] pixels = new int[width * height];
+    bitmap.getPixels(pixels, 0, width, 0, 0, width, height);
+    return encode(pixels, width, height, componentX, componentY);
+  }
+
+  private static String encode(int[] pixels, int width, int height, int componentX, int componentY) {
+
+    if (componentX < 1 || componentX > 9 || componentY < 1 || componentY > 9) {
+      throw new IllegalArgumentException("Blur hash must have between 1 and 9 components");
+    }
+    if (width * height != pixels.length) {
+      throw new IllegalArgumentException("Width and height must match the pixels array");
+    }
+
+    double[][] factors = new double[componentX * componentY][3];
+    for (int j = 0; j < componentY; j++) {
+      for (int i = 0; i < componentX; i++) {
+        double normalisation = i == 0 && j == 0 ? 1 : 2;
+        applyBasisFunction(pixels, width, height,
+            normalisation, i, j,
+            factors, j * componentX + i);
+      }
+    }
+
+    char[] hash = new char[1 + 1 + 4 + 2 * (factors.length - 1)]; // size flag + max AC + DC + 2 * AC components
+
+    long sizeFlag = componentX - 1 + (componentY - 1) * 9;
+    Base83.encode(sizeFlag, 1, hash, 0);
+
+    double maximumValue;
+    if (factors.length > 1) {
+      double actualMaximumValue    = max(factors, 1, factors.length);
+      double quantisedMaximumValue = Math.floor(Math.max(0, Math.min(82, Math.floor(actualMaximumValue * 166 - 0.5))));
+      maximumValue = (quantisedMaximumValue + 1) / 166;
+      Base83.encode(Math.round(quantisedMaximumValue), 1, hash, 1);
+    } else {
+      maximumValue = 1;
+      Base83.encode(0, 1, hash, 1);
+    }
+
+    double[] dc = factors[0];
+    Base83.encode(encodeDC(dc), 4, hash, 2);
+
+    for (int i = 1; i < factors.length; i++) {
+      Base83.encode(encodeAC(factors[i], maximumValue), 2, hash, 6 + 2 * (i - 1));
+    }
+    return new String(hash);
+  }
+
+  private static void applyBasisFunction(int[] pixels, int width, int height,
+                                         double normalisation, int i, int j,
+                                         double[][] factors, int index)
+  {
+    double r = 0, g = 0, b = 0;
+    for (int x = 0; x < width; x++) {
+      for (int y = 0; y < height; y++) {
+        double basis = normalisation
+                     * Math.cos((Math.PI * i * x) / width)
+                     * Math.cos((Math.PI * j * y) / height);
+        int pixel = pixels[y * width + x];
+        r += basis * sRGBToLinear((pixel >> 16) & 0xff);
+        g += basis * sRGBToLinear((pixel >> 8)  & 0xff);
+        b += basis * sRGBToLinear( pixel        & 0xff);
+      }
+    }
+    double scale = 1.0 / (width * height);
+    factors[index][0] = r * scale;
+    factors[index][1] = g * scale;
+    factors[index][2] = b * scale;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The result of `Integer` multiplication is cast into a larger type. If you think the result of the multiplication could exceed the size of an `Integer`, consider casting the arguments of the multiplication to avoid arithmetic overflows. For more information, see [CWE-190](https://cwe.mitre.org/data/definitions/190.html).